rosbag2_cpp/src/rosbag2_cpp/typesupport_helpers.cpp

```diff
@@ -105,33 +105,51 @@ get_typesupport(const std::string & type, const std::string & typesupport_identi
   std::string type_name;
   std::tie(package_name, middle_module, type_name) = extract_type_identifier(type);
 
-  std::string poco_dynamic_loading_error = "Something went wrong loading the typesupport library "
+  std::string rcutils_dynamic_loading_error =
+    "Something went wrong loading the typesupport library "
     "for message type " + package_name + "/" + type_name + ".";
 
   auto library_path = get_typesupport_library_path(package_name, typesupport_identifier);
 
-  try {
-    auto typesupport_library = std::make_shared<Poco::SharedLibrary>(library_path);
-
-    auto symbol_name = typesupport_identifier + "__get_message_type_support_handle__" +
-      package_name + "__" + (middle_module.empty() ? "msg" : middle_module) + "__" + type_name;
-
-    if (!typesupport_library->hasSymbol(symbol_name)) {
-      throw std::runtime_error(poco_dynamic_loading_error + " Symbol not found.");
+  rcutils_shared_library_t * typesupport_library = nullptr;
+
+  typesupport_library = new rcutils_shared_library_t;
+  *typesupport_library = rcutils_get_zero_initialized_shared_library();
+
+  rcutils_ret_t ret = rcutils_load_shared_library(typesupport_library, library_path.c_str());
+  if (ret != RCUTILS_RET_OK) {
+    delete typesupport_library;
+    if (ret == RCUTILS_RET_BAD_ALLOC) {
+      throw std::runtime_error(
+              std::string(rcutils_dynamic_loading_error + " failed to allocate memory"));
+    } else if (ret == RCUTILS_RET_INVALID_ARGUMENT) {
+      throw std::runtime_error(
+              std::string(rcutils_dynamic_loading_error + " invalid arguments"));
+    } else {
+      throw std::runtime_error(
+              std::string(
+                rcutils_dynamic_loading_error + " Library could not be found " +
+                library_path));
     }
+  }
 
-    const rosidl_message_type_support_t * (* get_ts)() = nullptr;
-    get_ts = (decltype(get_ts))typesupport_library->getSymbol(symbol_name);
-    auto type_support = get_ts();
+  auto symbol_name = typesupport_identifier + "__get_message_type_support_handle__" +
+    package_name + "__" + (middle_module.empty() ? "msg" : middle_module) + "__" + type_name;
 
-    if (!type_support) {
-      throw std::runtime_error(poco_dynamic_loading_error + " Symbol of wrong type.");
-    }
+  if (!rcutils_get_symbol(typesupport_library, symbol_name.c_str())) {
+    delete typesupport_library;
+    throw std::runtime_error(rcutils_dynamic_loading_error + " Symbol not found.");
+  }
+
+  const rosidl_message_type_support_t * (* get_ts)() = nullptr;
+  get_ts = (decltype(get_ts))rcutils_get_symbol(typesupport_library, symbol_name.c_str());
```

Comment:
What is this trying to do?

Reply:
to be fair, this is identical to the preexisting code. It's declaring a function pointer and calling an rcl API to set it so that it can be called.

Reply:
Agreed, though I was wondering if there's a better way to go about it since we're improving this anyway.

Reply:
I will put some time on this and in a this memory leak:

```
6: =================================================================
6: ==20733==ERROR: LeakSanitizer: detected memory leaks
6: Direct leak of 100 byte(s) in 1 object(s) allocated from:
6:     #0 0x7f8b08620acb in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xeacb)
6:     #1 0x7f8b07e623d0 in __default_allocate (/home/ahcorde/ros2_foxy/install/rcutils/lib/librcutils.so+0x43d0)
6:     #2 0x7f8b07e69cd0 in rcutils_load_shared_library (/home/ahcorde/ros2_foxy/install/rcutils/lib/librcutils.so+0xbcd0)
6:     #3 0x563ece336c48 in rosbag2_cpp::get_typesupport(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/ahcorde/ros2_foxy/build/rosbag2_cpp/test_ros2_message+0x66c48)
6:     #4 0x563ece2f2133 in Ros2MessageTest::get_allocated_message(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/ahcorde/ros2_foxy/build/rosbag2_cpp/test_ros2_message+0x22133)
6:     #5 0x563ece2ed467 in Ros2MessageTest_allocate_ros2_message_allocates_strings_message_with_empty_string_Test::TestBody() (/home/ahcorde/ros2_foxy/build/rosbag2_cpp/test_ros2_message+0x1d467)
6:     #6 0x563ece369143 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (/home/ahcorde/ros2_foxy/build/rosbag2_cpp/test_ros2_message+0x99143)
6:     #7 0x563ece362eee in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (/home/ahcorde/ros2_foxy/build/rosbag2_cpp/test_ros2_message+0x92eee)
 ```

Reply:
The unique_ptr is instantiated inside the function `get_typesupport` 

https://github.com/ros2/rosbag2/blob/d2cdcc3a0fef91fbaefac6398fc1150a5e4e0cab/rosbag2_cpp/src/rosbag2_cpp/typesupport_helpers.cpp#L116-L117

but we can't free this memory because we need to use the symbol outside the function and the library needs to be open. That's why the leak is happening.

any thoughts about how can we workaround this ?

Reply:
Interesting, maybe make this a `shared_ptr` then?

Reply:
The existing code was using a shared ptr that would have been freed at the end of that same function - did this leak already exist, or is this a difference in behavior between poco and the rcutils loader API?

Reply:
given that we instantiate a c-structure in the unique pointer, I would assume we'd need a custom deleter for the unique pointer which calls the appropriate `_fini` function for it.

I could imagine that the previous Poco instance had a properly defined destructor for itself in which case the default deleter of the unique pointer was sufficient.